src/main/java/minijava/ir/assembler/AssemblerGenerator.java

```diff
@@ -0,0 +1,584 @@
+      // at the start of the block that the phi is part of
+      CodeBlock phisBlock = getCodeBlockForNode(node);
+      Register intermediateReg = Register.EAX;
+      // we have to use a register to copy the value between two locations
+      phisBlock.addBlockStartInstruction(new Mov(tmpLocation, intermediateReg));
+      phisBlock.addBlockStartInstruction(new Mov(intermediateReg, res));
+    }
+    Block block = (Block) node.getBlock();
+    for (int i = 0; i < block.getPredCount(); i++) {
+      // the i.th block belongs to the i.th input edge of the phi node
+      Node inputNode = node.getPred(i);
+      // we get the correct block be taking the predecessors of the block of the phi node
+      Block inputBlock = (Block) block.getPred(i).getBlock();
+      CodeBlock inputCodeBlock = getCodeBlock(inputBlock);
+      Argument arg = allocator.getAsArgument(inputNode);
+      // we store the argument in a register
+      Register intermediateReg = Register.EAX;
+      inputCodeBlock.addPhiHelperInstruction(new Mov(arg, intermediateReg).firm(inputNode));
+      if (isPhiProneToLostCopies(node)) {
+        // if this phi is prone to lost copies then we copy the value only
+        // into the temporary variable
+        inputCodeBlock.addPhiHelperInstruction(new Mov(intermediateReg, tmpLocation));
+      } else {
+        // we copy it into Phis location
+        inputCodeBlock.addPhiHelperInstruction(new Mov(intermediateReg, res).firm(node));
+      }
+    }
+  }
+
+  @Override
+  public void visit(Load node) {
+    // firm graph:
+    // [Ptr node, might be a calculation!] <- [Load] <- [Proj res]
+    Argument arg = allocator.getAsArgument(node.getPtr());
+    Location res = allocator.getResultLocation(node);
+    CodeBlock block = getCodeBlockForNode(node);
+    Register intermediateReg = Register.EAX;
+    // load the pointer in a register
+    block.add(new Mov(arg, intermediateReg).firm(node.getPtr()));
+    // copy the value the pointer points to in a register
+    block.add(new Mov(new StackLocation(intermediateReg, 0), intermediateReg).firm(node));
+    // store the result
+    block.add(new Mov(intermediateReg, res));
+  }
+
+  @Override
+  public void visit(Store node) {
+    // firm graph:
+    // [Proj Mem] <---|
+    // [Ptr node] <---|
+    // [Value node] <-|Store] <- [Proj MM]
+    CodeBlock block = getCodeBlockForNode(node);
+    Register intermediateValReg = Register.EAX;
+    Register intermediateDestReg = Register.EBX;
+    Argument dest = allocator.getAsArgument(node.getPtr());
+    Argument newValue = allocator.getAsArgument(node.getValue());
+    // copy the new value into a register
+    block.add(new Mov(newValue, intermediateValReg).firm(node.getValue()));
+    // copy the address into another register
+    block.add(new Mov(dest, intermediateDestReg).firm(node.getPtr()));
+    // store the new value at its new location
+    block.add(new Mov(intermediateValReg, new StackLocation(intermediateDestReg, 0)).firm(node));
```

Comment:
`StackLocation` seems like kind of a misnomer if we also use it for memory access. Maybe rename it to `MemoryLocation` (maybe even without Location)?